src/app/SafeAttributePersistenceProvider.h

```diff
@@ -0,0 +1,253 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/ConcreteAttributePath.h>
+#include <app/data-model/Nullable.h>
+#include <app/util/attribute-metadata.h>
+#include <cstring>
+#include <inttypes.h>
+#include <lib/support/BufferReader.h>
+#include <lib/support/BufferWriter.h>
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * Interface for persisting attribute values. This will always write attributes in storage as little-endian
+ * and uses a different key to the AttributePersistenceProvider.
+ */
+
+class SafeAttributePersistenceProvider
+{
+public:
+    virtual ~SafeAttributePersistenceProvider() = default;
+    SafeAttributePersistenceProvider()          = default;
+
+    /**
+     * Write an attribute value from the attribute store (i.e. not a struct or
+     * list) to non-volatile memory.
+     *
+     * @param [in] aPath the attribute path for the data being written.
+     * @param [in] aValue the data to write.  Integers and floats are
+     *             represented in native endianness.  Strings are represented
+     *             as Pascal-style strings, as in ZCL, with a length prefix
+     *             whose size depends on the actual string type.  The length is
+     *             stored as little-endian.
+     *
+     *             Integer and float values have a size that matches the `size`
+     *             member of aMetadata.
+     *
+     *             String values have a size that corresponds to the actual size
+     *             of the data in the string (including the length prefix),
+     *             which is no larger than the `size` member of aMetadata.
+     */
+    virtual CHIP_ERROR SafeWriteValue(const ConcreteAttributePath & aPath, const ByteSpan & aValue) = 0;
+
+    /**
+     * Read an attribute value from non-volatile memory.
+     *
+     * @param [in]     aPath the attribute path for the data being persisted.
+     * @param [in]     aType the attribute type.
+     * @param [in]     aSize the attribute size.
```

Comment:
We should think about this one a little bit.  Fundamentally, we don't actually know the attribute type in this API.  I know the shared impl kind of wants to know it, but that's an implementation detail of the DefaultAttributePersistenceProvider and all it _really_ wants to know is a string/long-string/other tri-state.  It could handle that itself, by doing whatever hack around types it wants, or changing InternalReadValue to in fact take that tri-state or something.  So in this API ideally the type bit should not exist.

As for the size...  The way _this_ API uses size, the size always matches aValue.size(), no?  So again, we could have the hook just take the path and MutableByteSpan, and if DefaultAttributePersistenceProvider needs a size for its shared helper it can use aValue.size() for the size.  Safety will not suffer, because RawReadLowLevelBeCareful does its own size checks to make sure we got enough data, right?

Reply:
The size is has been removed.

Regarding the type, Things are a bit more complicated. I have found the `emberAfLocateAttributeMetadata` method which returns a pointer to an attribute's metadata from the path but this relies on the `gen_config.h` generated file. I feel like this code should not be dependant on generated files. Also, this will make it difficult to unit test as we would need to also have and setup a zap configuration.

Thinking about it, since the `SafeWriteValue` and `SafeReadValue` methods are protected, the `SafeAttributePersistanceProvider` can only be use for scalar values. So if we document this, the `DefaultAttributePersistanceProvider` has all the information it needs already.

See latest update.